### PR TITLE
Get encrypted token vectors

### PR DESCRIPTION
**Pull request by Nilanshrajput**
# Pull Request Template

## Description

Enabled transfer of encrypted vectors of individual tokens from remote worker.
call get_encrypted_token_vectors() which returns a  SMPC-encrypted tensor representing the 2D array of all vectors in the document this pointer points to.

## Type of change

Please mark options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test A : test_doc/test_get_token_vectors()

## Checklist:

* [x] I did follow the [contribution guidelines](https://github.com/OpenMined/PySyft/blob/master/CONTRIBUTING.md)
* [x] I have commented my code following [Google style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).
* [x] I have added tests for my changes

### CODE REVIEW

File: syfertext/doc.py [hunk 1]
```diff
@@ -211,6 +211,51 @@ def get_vector(self, excluded_tokens: Dict[str, Set[object]] = None):
             doc_vector = vectors / vector_count
         return doc_vector
 
+    def get_token_vectors(self, excluded_tokens: Dict[str, Set[object]] = None):
```
Comment (AlanAboudib) [inline syfertext/doc.py:214]: please add the type hint of the return type

File: syfertext/doc.py [hunk 2]
```diff
@@ -211,6 +211,51 @@ def get_vector(self, excluded_tokens: Dict[str, Set[object]] = None):
...
+        """Get the Numpy array of all the vectors corresponding to the tokens in the `Doc`,
+        excluding token according to the excluded_tokens dictionary.
+
+        Args
+            excluded_tokens (Dict): A dictionary used to ignore tokens of the document based on values
+                of their attributes.
+                Example: {'attribute1_name' : {value1, value2},'attribute2_name': {v1, v2}, ....}
```
Comment (AlanAboudib) [inline syfertext/doc.py:221]: Please add a space here {value1, value2},[ ]'attaribute2_name'....

File: syfertext/doc.py [hunk 3]
```diff
@@ -247,3 +293,40 @@ def get_encrypted_vector(
...
 
         return doc_vector
+
+    def get_encrypted_token_vectors(
+        self,
+        *workers: BaseWorker,
+        crypto_provider: BaseWorker = None,
+        requires_grad: bool = True,
+        excluded_tokens: Dict[str, Set[object]] = None,
+    ):
```
Comment (AlanAboudib) [inline syfertext/doc.py:303]: Please add the type hint of the return type like 'torch.Tensor'

File: syfertext/doc.py [hunk 4]
```diff
@@ -247,3 +293,40 @@ def get_encrypted_vector(
...
+            requires_grad (bool): A boolean flag indicating whether gradients are required or not.
+            excluded_tokens (Dict): A dictionary used to ignore tokens of the document based on values
+                of their attributes, the keys are the attributes names and they index, for efficiency, sets of values.
+                Example: {'attribute1_name' : {value1, value2},'attribute2_name': {v1, v2}, ....}
+
+        Returns:
+            Tensor: A SMPC-encrypted tensor representing the array of all vectors in this document, ingonoring the excluded token.
+        """
+
+        "You need at least two workers in order to encrypt the vector with SMPC"
```
Comment (AlanAboudib) [inline syfertext/doc.py:318]: Can you make this a comment starting with '#' instead of a python string

File: syfertext/pointers/doc_pointer.py [hunk 5]
```diff
@@ -57,6 +57,35 @@ def get_encrypted_vector(self, *workers, crypto_provider=None, requires_grad=Tru
 
         return doc_vector
 
+    def get_encrypted_token_vectors(self, *workers, crypto_provider=None, requires_grad=True):
+        """Get the mean of the vectors of each Token in this documents.
+
+        Args:
+            self (DocPointer): current pointer to a remote document.
```
Comment (AlanAboudib) [inline syfertext/pointers/doc_pointer.py:64]: No need to include self in the docstring

File: syfertext/pointers/doc_pointer.py [hunk 6]
```diff
@@ -57,6 +57,35 @@ def get_encrypted_vector(self, *workers, crypto_provider=None, requires_grad=Tru
...
+            workers (sequence of BaseWorker): A sequence of remote workers from .
+            crypto_provider (BaseWorker): A remote worker responsible for providing cryptography (SMPC encryption) functionalities.
```
Comment (AlanAboudib) [inline syfertext/pointers/doc_pointer.py:66]: please wrap long lines here and in all of the code you wrote

File: syfertext/pointers/doc_pointer.py [hunk 7]
```diff
@@ -57,6 +57,35 @@ def get_encrypted_vector(self, *workers, crypto_provider=None, requires_grad=Tru
 
         return doc_vector
 
+    def get_encrypted_token_vectors(self, *workers, crypto_provider=None, requires_grad=True):
```
Comment (AlanAboudib) [inline syfertext/pointers/doc_pointer.py:60]: You forgot `excluded_tokens`

File: tests/local_mode/test_doc.py [hunk 8]
```diff
@@ -130,3 +130,23 @@ def test_exclude_tokens_on_attr_values_doc():
...
     assert any(doc.get_vector() != doc_excluding_tokens.get_vector())
+
+
+def test_get_token_vectors():
+    """Test the get_token_vectors method"""
+
+    doc = nlp("Joey never ever share food")
+    doc_excluding_tokens = nlp("Joey never share food")
+
+    # add custom_attr to the last token, the word ever
```
Comment (AlanAboudib) [inline tests/local_mode/test_doc.py:141]: the word 'ever'

Comment (Nilanshrajput) [reply tests/local_mode/test_doc.py:141]: It's the difference between the 2 sentences(one exclude ever to test excludedd tokens are ignored)

Comment (AlanAboudib) [reply tests/local_mode/test_doc.py:141]: I know, I meant just add the quotes 'ever' instead of ever

Comment (Nilanshrajput) [reply tests/local_mode/test_doc.py:141]: actually behaviour will be changed due to tokenization

**Comment by Nilanshrajput** [comment]
changes were not affected, as i changed the position of head aster the update due to a bad merge, and didn't noticed the changes were gone

File: tests/local_mode/test_doc.py (hunk 8 above)
Comment (Nilanshrajput) [reply tests/local_mode/test_doc.py]: my bad, done :)

**Pull request merged**